pkg/servicemanager/manager.go

```diff
@@ -0,0 +1,177 @@
+package servicemanager
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/openshift/microshift/pkg/util/sigchannel"
+	"github.com/sirupsen/logrus"
+)
+
+type ServiceManager struct {
+	name string
+	deps []string
+
+	services   []Service
+	serviceMap map[string]Service
+}
+
+func NewServiceManager() *ServiceManager {
+	return &ServiceManager{
+		name: "service-manager",
+		deps: []string{},
+
+		services:   []Service{},
+		serviceMap: make(map[string]Service),
+	}
+}
+func (s *ServiceManager) Name() string           { return s.name }
+func (s *ServiceManager) Dependencies() []string { return s.deps }
+
+func (m *ServiceManager) AddService(s Service) error {
+	if _, exists := m.serviceMap[s.Name()]; exists {
+		return fmt.Errorf("service '%s' added more than once", s.Name())
+	}
+	for _, dependency := range s.Dependencies() {
+		// Enforce that services can only be added after adding their dependencies,
+		// i.e. they'll always remain topology sorted. Should we want to relax this
+		// constraint later, we can add topo sorting in the Run() step.
+		if _, exists := m.serviceMap[dependency]; !exists {
+			return fmt.Errorf("dependecy '%s' of service '%s' not yet defined", dependency, s.Name())
+		}
+	}
+
+	m.services = append(m.services, s)
```

Comment:
need a mutex here

Reply:
Thanks for the review, @rootfs. You mean for hedging against `AddService()` being called concurrently?
Note that at least currently, the order of adding services is significant (to enforce the invariant of dependencies being added before dependents), so `AddService()` shouldn't be called concurrently. Otherwise, we'd have to enable topo sorting. See also the comments in `AddService()` and `Run()`.

How should we proceed? Add mutex proactively or add note in method description?